### PR TITLE
docs: loss

### PR DESCRIPTION
**Pull request by bwanglzu**
Add loss function page for docs refactoring.

To review: https://ft-docs-loss--jina-docs.netlify.app/docs-loss/concepts/loss-functions/

---

- [x] This PR references an open issue
- [x] I have added a line about this change to CHANGELOG

### CODE REVIEW

File: docs/concepts/loss-functions.md [hunk 1]
```diff
@@ -1,6 +1,164 @@
...
+| query-article-irrelevant_article triplets | `MarginMSELoss`                                   |
+
+For more information, please checkout {doc}`/concepts/data-preparation` .
+
+## TripletMarginLoss
+
+`TripletMarginLoss` is a *contrastive* loss function, meaning that the loss is calculated by comparing the embeddings of multiple documents (3 to be exact) documents to each other.
+Each triplet of documents consists of an anchor document, a positive document and a negative document.
+The anchor and the positive document belong to the same class, and the negative document belongs to a different class.
+The goal of `TripletMarginLoss` is to maximise the difference between the distance from the anchor to the positive document, and the distance from the anchor to the negative document.
```
Comment (scott-martens) [inline docs/concepts/loss-functions.md:21]:
```suggestion
The goal of `TripletMarginLoss` is to minimize the difference between the distance from the anchor to the positive document, and maximize the distance from the anchor to the negative document.
```
 Check if I have the direction wrong.

File: docs/concepts/loss-functions.md [hunk 2]
```diff
@@ -1,6 +1,164 @@
...
+
+## ArcFaceLoss and CosFaceLoss
+
+SphereFace loss is a loss function that was first formulated for computer vision and face recognition tasks.
+Finetuner supports two variations of this loss function, `ArcFaceLoss` and `CosFaceLoss`.
+Instead of attempting to minimise the distance between positive pairs and maximise the distance between negative pairs, the SphereFace loss functions compare each sample with an estimate of the center point of each classes' embeddings.
```
Comment (scott-martens) [inline docs/concepts/loss-functions.md:27]:
```suggestion
Instead of attempting to minimize the distance between positive pairs and maximize the distance between negative pairs, the SphereFace loss functions compare each sample with an estimate of the center point of each class's embeddings.
```

File: docs/concepts/loss-functions.md [hunk 3]
```diff
@@ -1,6 +1,164 @@
...
+and attempt to minimize the *angular distance* between the document and its class centroid, and maximise the angular distance between the document and the centroids of the other classes.
```
Comment (scott-martens) [inline docs/concepts/loss-functions.md:28]:
```suggestion
and attempt to minimize the *angular distance* between the document and its class centroid, and maximize the angular distance between the document and the centroids of the other classes.
```

File: docs/concepts/loss-functions.md [hunk 4]
```diff
@@ -1,6 +1,164 @@
...
+If an item's embedding is within the boundary of the class it belongs to, then no loss is incurred. Choosing appropriate values for the margin and scaling parameter is very important for effective training.
+For more information on how `ArcFaceLoss` and `CosFaceLoss` calculate loss, and how these parameters affect the output, see this article on [loss metrics for deep learning](https://hav4ik.github.io/articles/deep-metric-learning-survey#cosface).  
+
+`TripletMarginLoss` uses a `ClassSampler` to construct batches with an equal number of samples of each class in the batch. However, since only one sample is needed to calculate the loss with the `ArcFaceLoss` and `CosFaceLoss` functions, there are no constraints on what each batch needs to contain.
+Therefore we can construct batches using random sampling, which is a much simpler and less time consuming method.
+By default, runs created using `ArcFaceLoss` or `CosfaceLoss` will use random sampling, however you can specify which type of sampling method you would like to use like so:
```
Comment (scott-martens) [inline docs/concepts/loss-functions.md:36]:
```suggestion
By default, runs created using `ArcFaceLoss` or `CosfaceLoss` will use random sampling, however you can specify which type of sampling method you would like to use with the `sampler` parameter:
```

File: docs/concepts/loss-functions.md [hunk 5]
```diff
@@ -1,6 +1,164 @@
...
+    loss='ArcFaceLoss',
++   sampler='random'      # use random sampling
++   sampler='class'       # use class sampling
++   sampler='auto'        # infer sampling method based on the loss function (default)
+
+)
+```
+
+In cases where the chosen loss function is a form of contrastive loss, such as the default `TripletMarginLoss`, or the `ClipLoss` function (the loss function used for `text-to-image` tasks), a class sampler is needed to properly function.
+In these cases, this `sampler` parameters is ignored and the `ClassSampler` is always used.
```
Comment (scott-martens) [inline docs/concepts/loss-functions.md:50]:
```suggestion
In these cases, this `sampler` parameter is ignored and the `ClassSampler` is always used.
```

File: docs/concepts/loss-functions.md [hunk 6]
```diff
@@ -1,6 +1,164 @@
...
+while minimizing the agreement between pairs that are unrelated.
+This encourages the model to learn a joint representation space where images and texts with similar meanings are close together.
+The loss function measures how well the CLIP model can encode the similarity between an image and its associated text.
+It uses a measure called cosine similarity,
+which calculates the cosine of the angle between two vectors.
+The higher the cosine similarity value, the closer the vectors are in the joint representation space.
+
+## CosineSimilarityLoss
+
+`CosineSimilarityLoss` is a regression loss function,
```
Comment (scott-martens) [inline docs/concepts/loss-functions.md:64]: What does regression have to do with it?

File: docs/concepts/loss-functions.md [hunk 7]
```diff
@@ -1,6 +1,164 @@
...
+which is calculated by comparing the cosine similarity of two embeddings against their ground-truth cosine similarity or some other numerical measure of similarity in the range of 0.0 (completely different) to 1.0 (identical). 
+The goal of `CosineSimilarityLoss` is to minimize the MSE (mean squared error) between document pair's cosine score and their ground-truth expected similarity, in order to optimize the model for semantic relatedness, i.e. between images and sentences that describe them in part, between documents that have similar content, etc.
+
+## MarginMSELoss
+
+```{admonition} Note
+:class: hint
+Use `MarginMSELoss` together with data synthesis job.
```
Comment (scott-martens) [inline docs/concepts/loss-functions.md:72]:
```suggestion
Use `MarginMSELoss` together with data synthesis.
```

A link for data synthesis here too.

File: docs/concepts/loss-functions.md [hunk 8]
```diff
@@ -1,6 +1,164 @@
...
+The margin is then obtained by taking the difference between these two similarity values.
+The loss is calculated as the mean squared error between the computed margin and a target margin value.
+
 ## Negative Mining
 
-should also be explained here
\ No newline at end of file
+Negative Mining is an advanced machine learning technique, which optimizes the way data is sampled from your training dataset.
+Usually, it aims at making the metric learning tasks for the model harder during the training. 
+In this way, it can lead to better fine-tuning results.
```
Comment (scott-martens) [inline docs/concepts/loss-functions.md:87]: This is counter-intuitive and might merit more explanation.

File: docs/concepts/loss-functions.md (hunk 6 above)
Comment (bwanglzu) [reply docs/concepts/loss-functions.md:64]: behind the scene CosineSimilarityLoss is minimising mean squared error @scott-martens

Comment (scott-martens) [reply docs/concepts/loss-functions.md:64]: Yeah okay.

File: docs/concepts/loss-functions.md [hunk 9]
```diff
@@ -1,6 +1,166 @@
...
+
 ## Negative Mining
 
-should also be explained here
\ No newline at end of file
+Negative Mining is an advanced machine learning technique, which optimizes the way data is sampled from your training dataset.
+Usually, it aims at making the metric learning tasks for the model harder during the training.
+During training, negative mining focuses on making the metric learning task more challenging for the model.
+It achieves this by carefully selecting negative samples from the dataset.
+In this way, it can lead to better fine-tuning results.
```
Comment (scott-martens) [inline docs/concepts/loss-functions.md:89]:
```suggestion
This can lead to better fine-tuning results.
```

File: docs/concepts/loss-functions.md [hunk 10]
```diff
@@ -1,6 +1,166 @@
...
+Metric Learning algorithms attempt to teach neural network models to tell
+which objects are semantically/visually similar and which ones are not.
+
+For uni-modal fine-tuning tasks such as text-to-text, image-to-image, or mesh-to-mesh,
+Finetuner constructs training data in the following way:
+
+![batch-sample](../imgs/batch-sampling.png)
+
+Assume we have a list of Documents belonging to four classes: `1`, `2`, `3`, and `4`,
+Finetuner will evenly sample *X* items per class to make a batch *B* which is encoded by the model into a set of embeddings.
```
Comment (scott-martens) [inline docs/concepts/loss-functions.md:104]:
```suggestion
Finetuner will evenly sample *X* items per class to make a batch, which is then encoded by the model into a set of embeddings.
```

File: docs/concepts/loss-functions.md [hunk 11]
```diff
@@ -1,6 +1,166 @@
...
+The objective is to pull the embeddings of items that belong to the same class closer together in the embedding space,
+while pushing the embeddings of items which belong to different classes farther away from each other.
+
+![training](../imgs/metric-train.png)
+
+
+### The Triplet Margin Miner
+
+For some triplets, the pre-trained model already performs well, i.e.
+
```
Comment (scott-martens) [inline docs/concepts/loss-functions.md:119]: delete this blank line

File: docs/concepts/loss-functions.md [hunk 12]
```diff
@@ -1,6 +1,166 @@
...
++ `D(anchor, neg) < D(anchor, pos) `, then `neg` can be considered as a "hard negative" (`2₄ - H`).
++ `D(anchor, pos) < D(anchor, neg) < D(anchor, pos) + margin`, where `neg` is a little further from the `pos`, but within the margin, then `neg` can be considered as a "semi-hard negative" (`2₄ - S`).
++ `D(anchor, neg) > D(anchor, pos) + margin`, then `neg` can be considered as "easy negative" (`2₄ - E`).
+
+Training is more effective when using only **hard** and **semi-hard** negatives, given a reasonable margin value to distinguish them from **easy** triplets.
+
+### Doing Negative Mining in Finetuner
+
+Finetuner is compatible with the miners provided by the [PyTorch Metric Learning](https://kevinmusgrave.github.io/pytorch-metric-learning) framework.
+To select a specific miner, pass its name to the `fit` function, e.g., `AngularMiner`, `TripletMarginMiner`, ...
```
Comment (scott-martens) [inline docs/concepts/loss-functions.md:140]:
```suggestion
To select a specific miner, pass its name to the `fit` function, e.g., `AngularMiner`, `TripletMarginMiner`, etc.
```

File: docs/concepts/loss-functions.md [hunk 13]
```diff
@@ -1,6 +1,166 @@
...
+
+Please note that the miner has to be compatible with the loss function you selected.
+For instance, if you choose to train a model with the `TripleMarginLoss`, you can use the `TripletMarginMiner`.
```
Comment (scott-martens) [inline docs/concepts/loss-functions.md:143]:
```suggestion
For instance, if you choose to train a model with the `TripleMarginLoss`, you can only use the `TripletMarginMiner`.
```

File: docs/concepts/loss-functions.md [hunk 14]
```diff
@@ -1,6 +1,166 @@
...
+While without this miner, all possible triples with an anchor, a positive, and a negative candidate are used to calculate the loss, the miner reduces this set of triples.
+By default, the miner only selects triples with hard negatives where the distance between the positive and the negative example is inside a margin of `0.2`.
```
Comment (scott-martens) [inline docs/concepts/loss-functions.md:145]:
```suggestion
By default, the miner only selects hard negative triples where the distance between the positive and the negative example is inside a margin of `0.2`.
```

File: docs/concepts/loss-functions.md [hunk 15]
```diff
@@ -1,6 +1,166 @@
...
+    ...,
+    loss='TripleMarginLoss',
++   miner='TripletMarginMiner',
++   miner_options={'margin': 0.3, 'type_of_triplets': 'hard'}
+)
+```
+
+Possible choices for `type_of_triplets` are:
+
++ `easy`: Use all easy triplets - all triplets that do not violate the margin.
```
Comment (scott-martens) [inline docs/concepts/loss-functions.md:159]:
```suggestion
+ `easy`: Use all easy triplets where the margin is less than the amount specified.
```

File: docs/concepts/loss-functions.md [hunk 16]
```diff
@@ -1,6 +1,166 @@
...
++ `semihard`: Use semi-hard triplets, but not hard triplets, i.e. those where difference in distance is within the specified margin.
++ `hard`: Use only hard triplets - the negative is closer to the anchor than the positive.
++ `all`: Use `hard` and `semihard` triples - all but the `easy` triples
```
Comment (scott-martens) [inline docs/concepts/loss-functions.md:162]:
```suggestion
+ `all`: Use `hard` and `semihard` triples, but no`easy` triples
```

File: docs/concepts/loss-functions.md [hunk 17]
```diff
@@ -1,6 +1,166 @@
...
+
+Finetuner takes `TripleMarginLoss` as its default loss function with no negative mining.
```
Comment (scott-martens) [inline docs/concepts/loss-functions.md:164]:
```suggestion
`TripleMarginLoss` is Finetuner's default loss function with no negative mining.
```

**Review by scott-martens — APPROVED** [review]
LGTM
